### DISCUSSION TITLE
Multiple Web Browsers

### DISCUSSION THREAD

**Discussion by JanSamusQF** [Voltstro-Studios/UnityWebBrowser#292] [Q&A] [answered]
### What platform are you experiencing this issue on?

Windows

### What architecture is your platform?

64-Bit

### What version of UWB are you using?

Latest VoltstroUPM Release

### What Unity version are you running?

2022.3.11f1

### Describe what the issue you are experiencing is.

When using two Web Browsers the first one works fine while the second one throws the following error:

[Web Browser]: [Engine]: Uncaught exception occured in the entry point!
System.InvalidOperationException: Failed to initialize CEF runtime.
   at Xilium.CefGlue.CefRuntime.Initialize(CefMainArgs, CefSettings, CefApp, IntPtr)
   at UnityWebBrowser.Engine.Cef.Core.CefEngineControlsManager.Init(ClientControlsActions, EnginePopupManager)
   at UnityWebBrowser.Engine.Cef.Core.UwbCefEngineEntry.EntryPoint(LaunchArguments, String[])
   at VoltstroStudios.UnityWebBrowser.Engine.Shared.Core.EngineEntryPoint.<>c__DisplayClass15_0.<Main>b__19(LaunchArguments parsedArgs)
UnityEngine.Logger:LogError (string,object)
VoltstroStudios.UnityWebBrowser.Logging.DefaultUnityWebBrowserLogger:Error (object) (at ./Library/PackageCache/dev.voltstro.unitywebbrowser@2.1.1/Runtime/Logging/DefaultUnityWebBrowserLogger.cs:36)
VoltstroStudios.UnityWebBrowser.Logging.ProcessLogHandler:HandleOutputProcessLog (object,System.Diagnostics.DataReceivedEventArgs) (at ./Library/PackageCache/dev.voltstro.unitywebbrowser@2.1.1/Runtime/Logging/ProcessLogHandler.cs:54)
System.Threading._ThreadPoolWaitCallback:PerformWaitCallback ()

Is there anything i have to configure to allow for multiple instances?

### Any additional info you like to provide?

I've updated to the current CEF Versions.

**Comment by Voltstro** [comment]
Ensure your communication layer is operating on different ports, and that either you have cache disabled, or change the cache path via [`WebBrowserClient.CachePath`](https://projects.voltstro.dev/UnityWebBrowser/latest/api/voltstrostudios.unitywebbrowser.core.webbrowserclient/#VoltstroStudios_UnityWebBrowser_Core_WebBrowserClient_CachePath).

**Comment by JanSamusQF** [comment]
I am creating a separate TCPCommunicationLayer ScriptableObject for each Whiteboard with different ports (5555,5556) and (5557,5558)
```
private void InitializeCommunicationLayer()
{
  // Create and initialize the TCP communication layer
  var tcpCommunicationLayer = ScriptableObject.CreateInstance<TCPCommunicationLayer>();
  CheckPort();
  tcpCommunicationLayer.inPort = inPort;
  tcpCommunicationLayer.outPort = outPort;
  clientManager.browserClient.communicationLayer = tcpCommunicationLayer;
}
```
while it works for a single whiteboard, for a second one it always throws the [Web Browser]: [Engine]: Uncaught exception occured in the entry point Exception.
I set the Log Severity to Debug and saw that it eventhough i create the TCPLayer it outputs this log aswell:
```
[Web Browser]: [Engine]: No communication layer provided, using default TCP...
UnityEngine.Logger:Log (UnityEngine.LogType,string,object)
VoltstroStudios.UnityWebBrowser.Logging.DefaultUnityWebBrowserLogger:Debug (object) (at ./Library/PackageCache/dev.voltstro.unitywebbrowser@2.1.1/Runtime/Logging/DefaultUnityWebBrowserLogger.cs:26)
VoltstroStudios.UnityWebBrowser.Logging.ProcessLogHandler:HandleOutputProcessLog (object,System.Diagnostics.DataReceivedEventArgs) (at ./Library/PackageCache/dev.voltstro.unitywebbrowser@2.1.1/Runtime/Logging/ProcessLogHandler.cs:50)
System.Threading._ThreadPoolWaitCallback:PerformWaitCallback ()
```
I disabled Cache for both and upped the timeout to 7000 with no changes. Is there anything i am missing?

**Reply by Voltstro** [reply, marked as answer]
I had a look into it quickly, looks like CEF has changed it behavior since I last experimented with running two instance.

What I did to get it working was to enable cache, and change [`WebBrowserClient.CachePath`](https://projects.voltstro.dev/UnityWebBrowser/latest/api/voltstrostudios.unitywebbrowser.core.webbrowserclient/#VoltstroStudios_UnityWebBrowser_Core_WebBrowserClient_CachePath) to a random path in the temp dir before UWB starts.

**Reply by JanSamusQF** [reply]
Awesome, thank you! This solved it.
